### DISCUSSION TITLE
method of characteristics

### DISCUSSION THREAD

**Discussion by gdmcbain** [kinnala/scikit-fem#720] [Ideas]
> The method of characteristics is tailored to solve *unsteady advection equations*

(Ern & Guermond 2004, _Theory and Practice of Finite Elements_, &sect;6.3.3 &lsquo;The method of characteristics&rsquo;).

(This is a stub which I'll fill out shortly, but for now to refer to from the discussion of the evaluation matrix #593.  I expect it to be a heavy load on that or `probes` #572.)

**Comment by kinnala** [comment]
Should it be solved like this: 1) find the points that map to the quadrature points through the advection map, 2) find out which of them are outside of the domain, 3) evluate rest of them through `probes`.

Then we could have a separate function which checks if a point is inside the domain or not.

**Comment by gdmcbain** [comment]
Following the discussion in #646, in particular your [question](https://github.com/kinnala/scikit-fem/issues/646#issuecomment-886016345)

> Is the time step so large that the point can move over several elements?

which is called in the finite difference context the Courant number *u* &Delta; *t* / &Delta; *x* or CFL condition after Courant, Friedrich, & Lewy (1928, 1967), I'm thinking that (3) should use a modified version of probes that is already told which cell each point is in and so doesn't have to call `element_finder` as in

https://github.com/kinnala/scikit-fem/blob/d08af2a3f4607e28023656b1b6ad025b1a530aa1/skfem/assembly/basis/cell_basis.py#L146

To begin with, one might check using the local-to-canonical map whether the point is in the guessed cell. For those that aren't, try the adjacent cell. I haven't looked at the details of what the local-to-canonical map looks like for a point not in the cell to see whether that might guide which facet has been crossed and so which of the adjacent cells is to be taken as the next guess, but if there's something there, it be better to use it than to fall back on an exhaustive search of the neighbours.

The mesh's facet connectivity array already indicates which facets are on the boundary of the domain, via the sentinel value &minus;1.

https://github.com/kinnala/scikit-fem/blob/0855dae117728588f269d968c8da4dbe06446d1b/skfem/mesh/mesh.py#L670

   —

 * Courant, R., Friedrichs, K. & Lewy, H. (1967). On the Partial Difference Equations of Mathematical Physics. *IBM Journal*, 11, 215--234.

**Comment by gdmcbain** [comment]
On the other hand, in examples like 
* exx 25, 28 (simple channels)
*  ex 27 (backwardfacing step)

it's easy to know which feet are upstream of the inlet because the mathematical description of the inlet is so simple (*x*=const.).

That shortcut might arise often enough to be worth exploiting. That would take care of 

> 2) find out which of them are outside of the domain

after which simple application of (3) to the remainder should be valid. (If not optimal, but of course premature optimization is to be avoided.)

**Comment by gdmcbain** [comment]
A very simple one-dimensional, uniform advecting velocity, zero diffusion, example is discussed in [&lsquo;Is this a good way to implement the method of characteristics for advection equation?&rsquo;](https://fenicsproject.org/qa/12744/this-implement-method-characteristics-advection-equation/) (_FEniCS Q&A_, 2017-03-19).

**Comment by kinnala** [comment]
There is also an efficient "in_polygon" implementation in matplotlib (Path class). There are also algorithms for 3D meshes, e.g. https://github.com/marmakoide/inside-3d-mesh
